File: RXSDK/RXSDK-Harmony-Native/HmsSdk/src/main/ets/types/BillingClient.ts
import { LoginDecorator } from "../base/LoginDecorator"
import Passport from "../base/Passport"
import ApiPath from "../constants/ApiPath"
import { RXRequest, RequestMethod } from "../net/RXRequest"
import { OrderData, PayParams, RXCallback, RXError, RXErrorCode, RXResult } from "./Index"

export abstract class BillingClient {
  protected _callbackFrom?: number

  public static init(_params?: string) {
  }

  @LoginDecorator
  async getOrder(params: PayParams): Promise<RXResult<OrderData>> {
    if (!params) {
      throw new RXError("params is null error", RXErrorCode.PAY_PARAMS_ERROR)
    }
    params.currency ??= "CNY"
    params.openid ??= Passport.openid
    params.age ??= Passport.loginData?.age
    this._callbackFrom = params.callback_from
    return RXRequest.request({
      path: ApiPath.ORDER,
      data: params,
      method: RequestMethod.POST,
    });
  }


  notifyResult(notifyUrl: string, data: Record<string, any>): Promise<RXResult> {
    if (!notifyUrl) {
      throw new RXError("notifyUrl is null error", RXErrorCode.PAY_PARAMS_ERROR)
    }
    let ret = RXRequest.request<RXResult>({
      path: notifyUrl,
      data: data,
      method: RequestMethod.POST,
    });
    return ret;
  }

  async pay(params: PayParams): Promise<RXResult> {
    let orderRet = await this.getOrder(this.handleParams(params))
    if (orderRet.code == 0) {
      return this.onOrderResponse(orderRet.data)
    } else {
      return orderRet
    }
  }

  handleParams(params: PayParams): PayParams {
    return params
  }

  queryPurchases(context) {
  }

  abstract name(): string

  abstract onOrderResponse(params: OrderData): Promise<RXResult>
}

File: RXSDK/RXSDK-Steam/Assets/RXSDK/Runtime/Billing/BillingClient.cs

using System;
using System.Collections.Generic;
using System.Linq;
using Newtonsoft.Json.Linq;
using RXSDK.Net;
using UnityEngine;

namespace RXSDK
{

    abstract class BillingClient
    {
        
        public abstract string BrandName { get; }
        public abstract void DoPay(OrderData orderData, RXCallback<object> callback, MonoBehaviour mono);
        public BillingCallbackType CallbackFrom { get; protected set; } = 0;
        public virtual void Order(MonoBehaviour mono, PayArgs args, RXCallback<OrderData> callback)
        {
            Order(mono, args.ToDictionary(), callback);
        }


        public virtual void Order(MonoBehaviour mono, Dictionary<string, object> keyValuePairs, RXCallback<OrderData> callback)
        {
            keyValuePairs ??= new Dictionary<string, object>();
            if (!keyValuePairs.ContainsKey("pay_type") && !keyValuePairs.ContainsKey("hq_type"))
            {
                keyValuePairs.Add("pay_type", BrandName);
            }
            string currency = "CNY";
            keyValuePairs.TryAdd("currency", currency);
            string openid = PassportManager.Instance.CurrentLoginData?.openid;
            keyValuePairs.TryAdd("openid", openid);

            keyValuePairs.TryAdd("age", PassportManager.Instance.CurrentLoginData?.age);

            if (keyValuePairs.ContainsKey("callback_from"))
            {
                CallbackFrom = ObjectUtility.To<BillingCallbackType>(keyValuePairs["callback_from"]);
            }

            RXWebRequest rXWebRequest = RXWebRequest.Create(APIPath.ORDER);
            rXWebRequest.SetPostData(keyValuePairs);
            rXWebRequest.PostAsync(mono, callback);
        }

        public virtual void Pay(MonoBehaviour mono, Dictionary<string, object> keyValuePairs, RXCallback<object> callback)
        {

            Order(mono, keyValuePairs, (ret, e) =>
            {
                var code = ret.code;
                var orderData = ret.data;
                var msg = ret.msg;
                if (code != 0)
                {
                    // callback?.Invoke(new RXResult<object> { code = code, data = orderData?.ToJson(), msg = msg });
                    RXUtility.InvokeCallback(callback, code, orderData?.ToJson(), msg);
                }
                else
                {
                    DoPay(orderData, callback, mono);
                    // Log.D("Application.platform:" + Application.platform);
                    // // callback?.Invoke(new RXResult<object> { code = code, data = orderData?.ToJson(), msg = msg });
                    // RXUtility.InvokeCallback(callback, code, orderData?.ToJson(), msg);

                }
            });
        }
        public virtual void RestorePurchase(MonoBehaviour mono, string notifyUrl = null, RXCallback<object> callback = null)
        {
            RXUtility.InvokeCallback(callback, (int)RXErrorCode.Success);

        }

        public virtual void VerifyPayment(MonoBehaviour mono, string notifyUrl, Dictionary<string, object> keyValuePairs, RXCallback<object> callback)
        {
            if (string.IsNullOrEmpty(notifyUrl))
            {
                RXUtility.InvokeCallback(callback, (int)RXErrorCode.PayError, null, "notify url null error");
                return;
            }

            keyValuePairs ??= new Dictionary<string, object>();
            RXWebRequest rXWebRequest = RXWebRequest.Create(notifyUrl);
            rXWebRequest.SetPostData(keyValuePairs);
            // rXWebRequest.Headers = SDKConfig.Instance.GetDefaultHeader();
            rXWebRequest.PostAsync<object>(mono, (ret, e) =>
            {
                var code = ret.code;
                var data = ret.data;
                var msg = ret.msg;
                if (code == 302402 || code == 302408 || code == 302409 || code == 302414)
                {
                    code = 0;
                }
                RXUtility.InvokeCallback(callback, code, data, msg);
            });
        }

    }
}